coralproject/talk
# Best Comments

* Story: https://www.pivotaltracker.com/n/projects/1863625/stories/138617379
* pre-implementation notes that @jde gave me goahead on. https://gist.github.com/gobengo/12cab171d4706509d345c8f33a71d9f2

## What does this PR do?

* Adds a button in the stream to tag comments as "BEST"
* Best comments get a tag label for all to see
* Fixes e2e tests

![best comment in action](http://i.imgur.com/LWpvi1o.gif)

code
* Add addCommentTag,removeCommentTag root mutations
* Add CommentsService#removeTag
* Test CommentsService#addTag and ensure same tag.name can't be added twice
* style changes to make sure that material-icons look good outside of the comment actions, multiple TagLabels are spaced reasonably, many comment actions can be in the footer on narrow screens without ugly wrapping
* Added class to allow selecting '.comment's only inside the Embed > Stream. And changed the embedStreamPage nightwatch element selectors to use it. This made e2e tests pass for me. See https://github.com/coralproject/talk/pull/334/files#r103414119

## How do I test this PR?

* Log in as a moderator and go to the embed
* Click the heart action
* the heart fills in, and a heart label is added at the top of the comment
* Refresh. Make sure everything persisted okay
* Click the filled in heart to toggle it off if you want (and refresh to verify), but then toggle back on
* Log out, the action to toggle best tag is gone because you're not a moderator. The <3 tag is still there for all to see though.

## Business Logic Changes

* Any MODERATOR or ADMIN can modify any allowed tag. This means, e.g. that a moderator can (by submitting graphql queries) remove the 'staff' tag from their comments. Is this okay? That seems both acceptable auth logic and, even if not, very low risk/harm that a moderator will ever do it.

## Questions For Product/Design

* [x] I added a class that would add `cursor: pointer` css to action buttons, but left it off of this PR for now. Can I style all the action buttons so the cursor changes to show it's clickable?
  * resolution: Sam said I could make all the action buttons cursor: pointer. So I did.
* [x] I used the heart icons instead of 'Tag as Best' (etc) copy because the latter was kind of long. What copy would you like? Or do you like the ❤️ ?
* [ ] Same for the indicator of best comment? Should it be a tag label like it is? Should it say "BEST" or something?
* [ ] Do you want any 'saving' indicator when a moderator clicks the BestButton? Right now I just disable the button while it's sending the request (to prevent double clicks). The slight change in style at least gives you a hint that something is happening.
* [x] What should happen if you click the BestButton and the request fails (e.g. your wifi goes out or the backend is down)? Right now there is no error message. It'll just behave like it didn't save.
  * resolution: Sam suggested I show the user a notification using the Notificatoin/SnackBar, so I wired that up.
